library/rmc/src/invariant.rs

```diff
@@ -0,0 +1,143 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0 OR MIT
+
+/// This module introduces the Invariant trait and as well as implementation for commonly used types.
+
+/// Types that implement a check to ensure its value is valid and safe to be used. See
+/// https://doc.rust-lang.org/stable/nomicon/what-unsafe-does.html for examples of valid values.
+///
+/// Implementations of Invariant traits must ensure that the current bit values of the given type
+/// is valid and that all its invariants hold.
+pub unsafe trait Invariant {
```

Comment:
That was a suggestion from @camshaft which I thought made a lot of sense. So I would prefer keeping it.
> 
> Is there going to be a marker trait for the safe function? If so that should be `unsafe` to implement (similar to [Send and Sync](https://doc.rust-lang.org/nomicon/safe-unsafe-meaning.html)).

Reply:
I was originally thinking it would be something like:

```rust
/// Types that implement this can be generated without any constraints or invariants
pub unsafe trait NondetTransmute { }

pub fn any<T: NondetTransmute>() -> T {
    unsafe { any_raw() }
}
```

You would then implement that for the primitives.

I'm trying to decide if the unsafe makes sense in its current form, though... On the one hand the `&self` argument is completely unvalidated and without knowing the RMC compiler internals, may result in UB when using it.

Is there a reason the trait method doesn't return an owned value instead? It would make the trait safe to implement and each implementation may or may not need to use unsafe to generate a value:

```rust
pub trait Arbitrary {
    fn any() -> Self;
}

impl Arbitrary for u8 {
    fn any() -> Self }
        unsafe { any_raw() }
    }
}
```

Reply:
Another concern is: in its current form, it's _very_ easy to miss validating a field when implementing the trait:

```rust
struct MyThing {
    field: char,
    other: char,
}

impl Invariant for MyThing {
    fn is_valid(& self) -> bool {
        self.field.is_valid() // missing `other`!
    }
}
```

Reply:
Thanks for clarifying. This is another possibility which uses a builder pattern. I thought the Invariant approach was a bit cleaner and easier to use. It also adds a method that can be used in assert statements which I thought was nice. I do agree that it has the downside of having a potential invalid `Self` and I think the unsafe impl is proper in this case.

I think they are somewhat equivalent though. For an arbitrary enum, it would go from:

```rust
unsafe impl Invariant for MyEnum {
    #[inline(always)]
    fn is_valid(&self) -> bool {
        matches!(*self, MyEnum::V1 | MyEnum::V2 | MyEnum::V3)
    }
}
```

to

```rust
impl Arbitrary for MyEnum {
    #[inline(always)]
    fn any() -> Self {
        match u8::any() {
            0 => MyEnum::V1,
            1 => MyEnum::V2,
            _ => MyEnum::V3,
        }
    }
}
```

Maybe we should check which one performs better or even offer both.

Reply:
I understand the concern, and I debated between the two options for a long time. Maybe we can have a quick sync tomorrow to discuss the options.

Another possibility is to add both methods to the type with default implementations. Something like:

```rust
pub unsafe trait Arbitrary {
    fn is_valid(&self) -> bool { true }
    fn any() -> Self {
        let s = unsafe { rmc::any_raw::<Self>() };
        rmc::assume(s.is_valid());
        s
    } 
}

pub fn any<T: Arbritary>() -> T {
    T::any()
}
```

Users can pick and choose which one to implement / use.